Bottom: fcf7bacf6643e43d50ddb9486d91b542f7877fd3
Top:    785e1396968ac458fd3125d445044474ab7f1e8e
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-04-21 16:19:21 +0100

Draw pin / pad names first, so they don't get masked by the pad

Since we use stencil masking to ensure we don't redraw an area
multiple times, confusingly, we need to draw the text first to
ensure it isn't occluded by the pad its-self.

HACK:

We probably draw the names multiple times - as we still use the
common draw.c code which assumes it can draw labels on top of
the pads.


---

diff --git a/src/hid/gtk/gtkhid-gl.c b/src/hid/gtk/gtkhid-gl.c
index deaff31..639b9bf 100644
--- a/src/hid/gtk/gtkhid-gl.c
+++ b/src/hid/gtk/gtkhid-gl.c
@@ -1073,6 +1073,16 @@ SetPVColor_inlayer (PinTypePtr Pin, LayerTypePtr Layer, int Type)
 
 
 static int
+pin_name_callback (const BoxType * b, void *cl)
+{
+  PinTypePtr pin = (PinTypePtr) b;
+
+  if (!TEST_FLAG (HOLEFLAG, pin) && TEST_FLAG (DISPLAYNAMEFLAG, pin))
+    DrawPinName (pin, 0);
+  return 1;
+}
+
+static int
 pin_inlayer_callback (const BoxType * b, void *cl)
 {
   SetPVColor_inlayer ((PinTypePtr) b, cl, PIN_TYPE);
@@ -1091,7 +1101,11 @@ via_inlayer_callback (const BoxType * b, void *cl)
 static int
 pin_callback (const BoxType * b, void *cl)
 {
-  DrawPlainPin ((PinTypePtr) b, false);
+  PinTypePtr pin = (PinTypePtr) b;
+
+  if (!TEST_FLAG (HOLEFLAG, pin) && TEST_FLAG (DISPLAYNAMEFLAG, pin))
+    DrawPinName (pin, 0);
+  DrawPlainPin (pin, false);
   return 1;
 }
 
@@ -1100,7 +1114,11 @@ pad_callback (const BoxType * b, void *cl)
 {
   PadTypePtr pad = (PadTypePtr) b;
   if (FRONT (pad))
+  {
+    if (TEST_FLAG (DISPLAYNAMEFLAG, pad))
+      DrawPadName (pad, 0);
     DrawPad (pad, 0);
+  }
   return 1;
 }
 
@@ -1332,6 +1350,9 @@ DrawLayerGroup (int group, const BoxType * screen)
 
       /* Draw pins, vias and pads on this layer */
       if (!global_view_2d && rv) {
+        if (PCB->PinOn &&
+            (group == solder_group || group == component_group))
+          r_search (PCB->Data->pin_tree, screen, NULL, pin_name_callback, Layer);
         if (PCB->PinOn) r_search (PCB->Data->pin_tree, screen, NULL, pin_inlayer_callback, Layer);
         if (PCB->ViaOn) r_search (PCB->Data->via_tree, screen, NULL, via_inlayer_callback, Layer);
         if ((group == component_group && !SWAP_IDENT) ||
